Deprecate Index.summary

```python
In [88]: idx = pd.Index([1, 2, 3])

In [89]: idx.summary()
Out[89]: 'Int64Index: 3 entries, 1 to 3'
```

Apologies if any of this is super naive. I'm new.

Deleting the method and any tests directly targeting it isn't enough. There are other failures like

```
__ TestDataFrameReprInfoEtc.test_info_duplicate_columns_shows_correct_dtypes ___

self = <pandas.tests.frame.test_repr_info.TestDataFrameReprInfoEtc object at 0x7f34597891d0>

    def test_info_duplicate_columns_shows_correct_dtypes(self):
        # GH11761
        io = StringIO()

        frame = DataFrame([[1, 2.0]],
                          columns=['a', 'a'])
        frame.info(buf=io)
        io.seek(0)
        lines = io.readlines()
>       assert 'a    1 non-null int64\n' == lines[3]
E       AssertionError: assert 'a    1 non-null int64\n' == 'a    1 non-null float64\n'
E         - a    1 non-null int64
E         ?                 ^^
E         + a    1 non-null float64
E         ?                 ^^^^
```

The reason is the output of `summary` is included in the output of `pandas.DataFrame.info`.

```
In [11]: # pretend the variables in the test are available here

In [12]: lines
Out[12]: 
["<class 'pandas.core.frame.DataFrame'>\n",
 'RangeIndex: 1 entries, 0 to 0\n',
 'Data columns (total 2 columns):\n',
 'a    1 non-null int64\n',
 'a    1 non-null float64\n',
 'dtypes: float64(1), int64(1)\n',
 'memory usage: 96.0 bytes\n']
```

(That second element is the output from `summary`.)

Would it be more appropriate to
* replace `Index.summary` with `Index.info`,
* make `info` more useful (per above comments), and
* update tests?

--- a/doc/source/whatsnew/v0.23.0.txt
+++ b/doc/source/whatsnew/v0.23.0.txt
@@ -736,6 +736,7 @@ Deprecations
 - :attr:`Timestamp.weekday_name`, :attr:`DatetimeIndex.weekday_name`, and :attr:`Series.dt.weekday_name` are deprecated in favor of :meth:`Timestamp.day_name`, :meth:`DatetimeIndex.day_name`, and :meth:`Series.dt.day_name` (:issue:`12806`)
 
 - ``pandas.tseries.plotting.tsplot`` is deprecated. Use :func:`Series.plot` instead (:issue:`18627`)
+- ``Index.summary()`` is deprecated and will be removed in a future version (:issue:`18217`)
 
 .. _whatsnew_0230.prior_deprecations:
 

--- a/pandas/core/frame.py
+++ b/pandas/core/frame.py
@@ -2005,7 +2005,7 @@ class DataFrame(NDFrame):
         lines = []
 
         lines.append(str(type(self)))
-        lines.append(self.index.summary())
+        lines.append(self.index._summary())
 
         if len(self.columns) == 0:
             lines.append('Empty %s' % type(self).__name__)
@@ -2341,7 +2341,7 @@ class DataFrame(NDFrame):
         try:
             if key in self.columns and not is_mi_columns:
                 return self._getitem_column(key)
-        except:
+        except Exception:
             pass
 
         # see if we can slice the rows
@@ -2843,7 +2843,7 @@ class DataFrame(NDFrame):
         if not len(self.index) and is_list_like(value):
             try:
                 value = Series(value)
-            except:
+            except Exception:
                 raise ValueError('Cannot set a frame with no defined index '
                                  'and a value that cannot be converted to a '
                                  'Series')
@@ -6828,7 +6828,7 @@ def _prep_ndarray(values, copy=True):
                 values = np.array([convert(v) for v in values])
             else:
                 values = convert(values)
-        except:
+        except Exception:
             values = convert(values)
 
     else:

--- a/pandas/core/frame.py
+++ b/pandas/core/frame.py
@@ -2341,7 +2341,7 @@ class DataFrame(NDFrame):
         try:
             if key in self.columns and not is_mi_columns:
                 return self._getitem_column(key)
-        except Exception:
+        except:
             pass
 
         # see if we can slice the rows
@@ -2843,7 +2843,7 @@ class DataFrame(NDFrame):
         if not len(self.index) and is_list_like(value):
             try:
                 value = Series(value)
-            except Exception:
+            except:
                 raise ValueError('Cannot set a frame with no defined index '
                                  'and a value that cannot be converted to a '
                                  'Series')
@@ -6828,7 +6828,7 @@ def _prep_ndarray(values, copy=True):
                 values = np.array([convert(v) for v in values])
             else:
                 values = convert(values)
-        except Exception:
+        except:
             values = convert(values)
 
     else:

--- a/pandas/core/indexes/base.py
+++ b/pandas/core/indexes/base.py
@@ -1388,8 +1388,15 @@ class Index(IndexOpsMixin, PandasObject):
     def _summary(self, name=None):
         """
         Return a summarized representation
-        .. deprecated:: 0.23.0
-           No longer supported
+
+        Parameters
+        ----------
+        name : str
+            name to use in the summary representation
+
+        Returns
+        -------
+        String with a summarized representation of the index
         """
         if len(self) > 0:
             head = self[0]

--- a/pandas/core/indexes/base.py
+++ b/pandas/core/indexes/base.py
@@ -1420,7 +1420,6 @@ class Index(IndexOpsMixin, PandasObject):
         """
         Return a summarized representation
         .. deprecated:: 0.23.0
-           No longer supported
         """
         warnings.warn("'summary' is deprecated and will be removed in a "
                       "future version.", FutureWarning, stacklevel=2)

--- a/pandas/core/indexes/base.py
+++ b/pandas/core/indexes/base.py
@@ -1385,7 +1385,7 @@ class Index(IndexOpsMixin, PandasObject):
         # to disable groupby tricks in MultiIndex
         return False
 
-    def summary(self, name=None):
+    def _summary(self, name=None):
         if len(self) > 0:
             head = self[0]
             if (hasattr(head, 'format') and
@@ -1403,6 +1403,15 @@ class Index(IndexOpsMixin, PandasObject):
         if name is None:
             name = type(self).__name__
         return '%s: %s entries%s' % (name, len(self), index_summary)
+
+    def summary(self, name=None):
+        """
+        .. deprecated:: 0.23.0
+           No longer supported
+        """
+        warnings.warn("'summary' is deprecated and will be removed in a "
+                      "future version.", FutureWarning, stacklevel=2)
+        return self._summary(name)
 
     def _mpl_repr(self):
         # how to represent ourselves to matplotlib

--- a/pandas/core/indexes/base.py
+++ b/pandas/core/indexes/base.py
@@ -1386,6 +1386,11 @@ class Index(IndexOpsMixin, PandasObject):
         return False
 
     def _summary(self, name=None):
+        """
+        Return a summarized representation
+        .. deprecated:: 0.23.0
+           No longer supported
+        """
         if len(self) > 0:
             head = self[0]
             if (hasattr(head, 'format') and
@@ -1406,6 +1411,7 @@ class Index(IndexOpsMixin, PandasObject):
 
     def summary(self, name=None):
         """
+        Return a summarized representation
         .. deprecated:: 0.23.0
            No longer supported
         """

--- a/pandas/core/indexes/datetimelike.py
+++ b/pandas/core/indexes/datetimelike.py
@@ -1049,7 +1049,7 @@ class DatetimeIndexOpsMixin(object):
         return self._shallow_copy(result,
                                   **self._get_attributes_dict())
 
-    def summary(self, name=None):
+    def _summary(self, name=None):
         """
         return a summarized representation
         """
@@ -1070,6 +1070,15 @@ class DatetimeIndexOpsMixin(object):
         # display as values, not quoted
         result = result.replace("'", "")
         return result
+
+    def summary(self, name=None):
+        """
+        .. deprecated:: 0.23.0
+           No longer supported
+        """
+        warnings.warn("'summary' is deprecated and will be removed in a "
+                      "future version.", FutureWarning, stacklevel=2)
+        return self._summary(name)
 
     def _concat_same_dtype(self, to_concat, name):
         """

--- a/pandas/core/indexes/datetimelike.py
+++ b/pandas/core/indexes/datetimelike.py
@@ -1051,7 +1051,9 @@ class DatetimeIndexOpsMixin(object):
 
     def _summary(self, name=None):
         """
-        return a summarized representation
+        Return a summarized representation
+        .. deprecated:: 0.23.0
+           No longer supported
         """
         formatter = self._formatter_func
         if len(self) > 0:
@@ -1070,15 +1072,6 @@ class DatetimeIndexOpsMixin(object):
         # display as values, not quoted
         result = result.replace("'", "")
         return result
-
-    def summary(self, name=None):
-        """
-        .. deprecated:: 0.23.0
-           No longer supported
-        """
-        warnings.warn("'summary' is deprecated and will be removed in a "
-                      "future version.", FutureWarning, stacklevel=2)
-        return self._summary(name)
 
     def _concat_same_dtype(self, to_concat, name):
         """

--- a/pandas/core/indexes/datetimelike.py
+++ b/pandas/core/indexes/datetimelike.py
@@ -1052,8 +1052,15 @@ class DatetimeIndexOpsMixin(object):
     def _summary(self, name=None):
         """
         Return a summarized representation
-        .. deprecated:: 0.23.0
-           No longer supported
+
+        Parameters
+        ----------
+        name : str
+            name to use in the summary representation
+
+        Returns
+        -------
+        String with a summarized representation of the index
         """
         formatter = self._formatter_func
         if len(self) > 0:

--- a/pandas/tests/indexes/datetimes/test_formats.py
+++ b/pandas/tests/indexes/datetimes/test_formats.py
@@ -182,7 +182,7 @@ class TestDatetimeIndexRendering(object):
 
         for idx, expected in zip([idx1, idx2, idx3, idx4, idx5, idx6],
                                  [exp1, exp2, exp3, exp4, exp5, exp6]):
-            result = idx.summary()
+            result = idx._summary()
             assert result == expected
 
     def test_dti_business_repr(self):
@@ -191,15 +191,15 @@ class TestDatetimeIndexRendering(object):
 
     def test_dti_business_summary(self):
         rng = pd.bdate_range(datetime(2009, 1, 1), datetime(2010, 1, 1))
-        rng.summary()
-        rng[2:2].summary()
+        rng._summary()
+        rng[2:2]._summary()
 
     def test_dti_business_summary_pytz(self):
-        pd.bdate_range('1/1/2005', '1/1/2009', tz=pytz.utc).summary()
+        pd.bdate_range('1/1/2005', '1/1/2009', tz=pytz.utc)._summary()
 
     def test_dti_business_summary_dateutil(self):
         pd.bdate_range('1/1/2005', '1/1/2009',
-                       tz=dateutil.tz.tzutc()).summary()
+                       tz=dateutil.tz.tzutc())._summary()
 
     def test_dti_custom_business_repr(self):
         # only really care that it works
@@ -209,12 +209,13 @@ class TestDatetimeIndexRendering(object):
     def test_dti_custom_business_summary(self):
         rng = pd.bdate_range(datetime(2009, 1, 1), datetime(2010, 1, 1),
                              freq='C')
-        rng.summary()
-        rng[2:2].summary()
+        rng._summary()
+        rng[2:2]._summary()
 
     def test_dti_custom_business_summary_pytz(self):
-        pd.bdate_range('1/1/2005', '1/1/2009', freq='C', tz=pytz.utc).summary()
+        pd.bdate_range('1/1/2005', '1/1/2009', freq='C',
+                       tz=pytz.utc)._summary()
 
     def test_dti_custom_business_summary_dateutil(self):
         pd.bdate_range('1/1/2005', '1/1/2009', freq='C',
-                       tz=dateutil.tz.tzutc()).summary()
+                       tz=dateutil.tz.tzutc())._summary()

--- a/pandas/tests/indexes/period/test_formats.py
+++ b/pandas/tests/indexes/period/test_formats.py
@@ -207,10 +207,3 @@ Freq: Q-DEC"""
                                   exp6, exp7, exp8, exp9]):
             result = idx._summary()
             assert result == expected
-
-    def test_summary_deprecated(self):
-        # GH18217
-        idx = PeriodIndex(['2011-01-01'], freq='D')
-
-        with tm.assert_produces_warning(FutureWarning):
-            idx.summary()

--- a/pandas/tests/indexes/period/test_formats.py
+++ b/pandas/tests/indexes/period/test_formats.py
@@ -205,5 +205,12 @@ Freq: Q-DEC"""
                                   idx6, idx7, idx8, idx9],
                                  [exp1, exp2, exp3, exp4, exp5,
                                   exp6, exp7, exp8, exp9]):
-            result = idx.summary()
+            result = idx._summary()
             assert result == expected
+
+    def test_summary_deprecated(self):
+        # GH18217
+        idx = PeriodIndex(['2011-01-01'], freq='D')
+
+        with tm.assert_produces_warning(FutureWarning):
+            idx.summary()

--- a/pandas/tests/indexes/test_base.py
+++ b/pandas/tests/indexes/test_base.py
@@ -1055,13 +1055,20 @@ class TestIndex(Base):
         assert not self.intIndex.is_all_dates
 
     def test_summary(self):
-        self._check_method_works(Index.summary)
+        self._check_method_works(Index._summary)
         # GH3869
         ind = Index(['{other}%s', "~:{range}:0"], name='A')
-        result = ind.summary()
+        result = ind._summary()
         # shouldn't be formatted accidentally.
         assert '~:{range}:0' in result
         assert '{other}%s' in result
+
+    # GH18217
+    def test_summary_deprecated(self):
+        ind = Index(['{other}%s', "~:{range}:0"], name='A')
+
+        with tm.assert_produces_warning(FutureWarning):
+            ind.summary()
 
     def test_format(self):
         self._check_method_works(Index.format)

--- a/pandas/tests/indexes/timedeltas/test_formats.py
+++ b/pandas/tests/indexes/timedeltas/test_formats.py
@@ -5,6 +5,8 @@ import pytest
 import pandas as pd
 from pandas import TimedeltaIndex
 
+import pandas.util.testing as tm
+
 
 class TestTimedeltaIndexRendering(object):
     @pytest.mark.parametrize('method', ['__repr__', '__unicode__', '__str__'])
@@ -67,6 +69,11 @@ class TestTimedeltaIndexRendering(object):
                 result = repr(pd.Series(idx))
                 assert result == expected
 
+    def test_summary_deprecated(self):
+        idx = TimedeltaIndex([], freq='D')
+        with tm.assert_produces_warning(FutureWarning):
+            idx.summary()
+
     def test_summary(self):
         # GH#9116
         idx1 = TimedeltaIndex([], freq='D')
@@ -92,5 +99,5 @@ class TestTimedeltaIndexRendering(object):
 
         for idx, expected in zip([idx1, idx2, idx3, idx4, idx5],
                                  [exp1, exp2, exp3, exp4, exp5]):
-            result = idx.summary()
+            result = idx._summary()
             assert result == expected

--- a/pandas/tests/indexes/timedeltas/test_formats.py
+++ b/pandas/tests/indexes/timedeltas/test_formats.py
@@ -5,8 +5,6 @@ import pytest
 import pandas as pd
 from pandas import TimedeltaIndex
 
-import pandas.util.testing as tm
-
 
 class TestTimedeltaIndexRendering(object):
     @pytest.mark.parametrize('method', ['__repr__', '__unicode__', '__str__'])
@@ -68,11 +66,6 @@ class TestTimedeltaIndexRendering(object):
                                      [exp1, exp2, exp3, exp4, exp5]):
                 result = repr(pd.Series(idx))
                 assert result == expected
-
-    def test_summary_deprecated(self):
-        idx = TimedeltaIndex([], freq='D')
-        with tm.assert_produces_warning(FutureWarning):
-            idx.summary()
 
     def test_summary(self):
         # GH#9116

